### DISCUSSION TITLE
Langium 3.0.0, "getOrCreateDocument"

### DISCUSSION THREAD

**Discussion by goto40** [eclipse-langium/langium#1406] [Q&A] [answered]
In Langium 2.1.3 I use `LangiumDocuments.getOrCreateDocument` to get a document synchronously.
Is there a way to do this in Langium 3.0.0 as well?

**Comment by spoenemann** [comment]
Yes, `getDocument` – but the downside is that this can return `undefined`. Creating an AST from text is now async by default.

**Reply by cdietrich** [reply]
seee also discussion here https://github.com/eclipse-langium/langium/pull/1306

**Reply by goto40** [reply]
I have rewritten my toolfunction (which used `getOrCreateDocument` as follows (in some special cases we need to load a set of given files):
```ts
export function loadDocFromUri(services: LangiumServices, uri: URI): LangiumDocument {
    if (!fs.existsSync(uri.fsPath)) {
        throw new Error(`File ${uri.fsPath} does not exist.`);
    }
    const result = services.shared.workspace.LangiumDocuments.getDocument(uri);
    if (result===undefined) {
        const fileContent = fs.readFileSync(uri.fsPath, 'utf-8');
        const doc = new DefaultLangiumDocumentFactory(services.shared).fromString(fileContent, uri);
        services.shared.workspace.LangiumDocuments.addDocument(doc);
        return doc;
    }
    else {
        return result;
    }
}
```

Does this look awkward somehow? Or is it a valid implementation to read a model file (as sync function).

**Reply by msujew** [reply, marked as answer]
Instead of `new DefaultLangiumDocumentFactory` you should just use `services.shared.workspace.LangiumDocumentFactory`, but otherwise, this seems reasonable. I'm wondering though, why you explicitly need this to be synchronous?

**Reply by goto40** [reply]
We have this for cases where we generate code for individual model files (and load all imports manually based on import statements). The code was synchronous and I did not want to rewrite the whole chain of calls...

We use this also for unittests.

Note: the normal use case is always to load models based on a workspace folder (like vscode does it). But we still support the other usage as well. This "manual" loading" also allows handle symlinks in our paths, which somehow fails in cases where we load a whole workspace folder (even that vscode is able to handle the symlinks). But symlinks are not a good usage to handle stuff anyway... (but people are using it sometimes).
